### ISSUE TITLE
Multiple lines auto added when more than one datagrid present

### ISSUE THREAD

**Issue by ewohnlich** [collective/collective.z3cform.datagridfield#96]
May or may not be related to https://github.com/collective/collective.z3cform.datagridfield/issues/82. I am on version 1.5.2, Plone 5.2, python 3. I am getting multiple rows added when auto completing the first row when I have multiple fields that use a datagrid widget. If it matters, I do see in the network tab that init_field.js is called again for each field that uses this widget.

I have tested with both 1.5.1 and 1.5.2 in Plone 5.2 and got the same behavior. Previously in Plone 5.1 we were on DGF version 1.3.0 and did NOT have the problem.

**Comment by ewohnlich** [comment]
Commenting out the <script> line for it in datagridfield_input.pt seems to solve this problem. I'm not sure what that file intends to solve.

**Comment by ewohnlich** [comment]
Is having multiple DGFs on one form supported/tested? I didn't see anything obviously doing that in the demo section. I've tried to eliminate any integration mistakes on my end as a factor :)

**Comment by ewohnlich** [comment]
Upon further testing this only seems to be an issue for migrated sites. New 5.2.1 sites do not have an issue, even on forms with multiple DGF fields. On migrated ones, most of our sites have been around since Plone 4 at least so it's difficult to determine where a problem is though I suspect the big change with plone.staticresources in 5.2 is a culprit.

I did see that I had a DGF related resource in the plone-legacy bundle, though oddly removing that did not seem to have an effect. I am not sure I have a good enough understanding of all of the caching related here to rule that out as the cause though. I do know that removing bundle with "default" resolved the issue on our migrated sites though. I don't believe I've missed an upgrade step.

I'll continue to test and update, hopefully I can provide a clearer answer for anyone else migrating old sites that use this.

**Comment by mauritsvanrees** [comment]
I see this problem too. Plone 5.2, Python 3. I see this on new sites. I have two content types, one of them with one datagrid field, the other with three.  On Plone 5.1 this works fine. On Plone 5.2.1 not.  The type with only one field works well. With three fields, when I fill in a column and tab to the next, I automatically get four extra rows, instead of just one. 

This is with `collective.z3cform.datagridfield` 1.5.2 (or current master). When I go back to 1.5.1, the problem is solved. So I guess this was caused by PR #93 trying to fix issue #92.

With the old code, `dataGridField2Functions.init` is called once, and `$(".datagridwidget-body")` matches once, even though with a javascript console I see more than one such object.  All works fine though.

With the new code, `dataGridField2Functions.init` is called multiple times. `$(".datagridwidget-body")` matches sometimes once, sometimes twice.

@ksuess since you did the PR, do you have an idea?

**Comment by mauritsvanrees** [comment]
cc @oggers since you added issue #92 and added a patch there that did not get applied in the final PR.

**Comment by ksuess** [comment]
@mauritsvanrees 
Yes, I can reproduce this. dataGridField2Functions.init is called once per widget, due to the fact that init_field.js is per widget. :(
So one option is to use a cookie to check if dataGridField2Functions.init is already called.
A revert of the bug fix is no option, the bug fix was made to ensure the init is done not before the DOM is ready for binding etc..

**Comment by mauritsvanrees** [comment]
Fixed in `collective.z3cform.datagridfield` 1.5.3.

**Comment by abosio** [comment]
I'm still observing this problem on a Plone 5.2 site migrated from 5.1.5. I've pinned DGF 1.5.3 and:
- Reimported its default profile
- Run its upgrade step
- Uninstalled and installed (in Add-ons)
With instance restarts in between each. I can see the bug fix code is present in init_field.js in the Debugger panel of Web Inspector and verified it is only getting called once with a console.log().

**Comment by PatrickDowney** [comment]
I just came across this issue and 1.5.3 still seems to .  Using Plone 5.2, Python 3, testing on new Plone instance.  I'm using it for a registry edit form.
Currently, the workaround seems to be submitting the form, and then when the page comes back with the validation error, delete the extra row, and then resubmit, and if the form already has a record, use '+' to add a new empty row instead of using the empty bottom row.  Perhaps, is it possible to disable it from automatically adding new rows?

**Comment by abosio** [comment]
@PatrickDowney I thought it was still doing it for a Plone 5.2 on Python 3 project of ours but it turned out to be a second copy of datagridfield’s JS getting loaded so that is something to check on.

**Comment by PatrickDowney** [comment]
Yes, that was the issue.  Sorry about that.  Thank you for letting me know and I appreciate the quick response.

**Comment by ksuess** [comment]
@PatrickDowney The bug was fixed by making sure that the initialization of a data grid field happens only once. If you still find more than one row auto appended then please have a look via developer tools for dataGridField2Functions.init(). This should be found in 
init_field.js:
```
      if ($("body").attr("class").indexOf("datagridfield-initialized") === -1) {
        $("body").addClass("datagridfield-initialized");
        window.dataGridField2Functions.init();
      };
```
If this is not the case, your datagridfield add-on is not up to date or you have some extra code that should be removed.

If the problem still exists, please file a new issue with a description of your case.

Independent of the problem, you can of course also disable the auto-append option. Have a look at
http://localhost:8083/Plone/@@demo-collective.z3cform.datagrid-menu
